addons/binding/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/PlugwiseThingDiscoveryService.java

```diff
@@ -0,0 +1,387 @@
+            discoveredNodes.put(macAddress, new DiscoveredNode(macAddress));
+            updateInformation(macAddress);
+        } else {
+            logger.debug("Already discovered node ({})", macAddress);
+        }
+    }
+
+    protected void discoverNodes() {
+        if (getStickStatus() != ThingStatus.ONLINE) {
+            logger.debug("Discovery with role call not possible (Stick status is {})", getStickStatus());
+        } else if (getCirclePlusMAC() == null) {
+            logger.debug("Discovery with role call not possible (Circle+ MAC address is null)");
+        } else if (currentRoleCall.isRoleCalling) {
+            logger.debug("Discovery with role call not possible (already role calling)");
+        } else {
+            stickHandler.addMessageListener(this);
+            discoveredNodes.clear();
+            currentRoleCall.isRoleCalling = true;
+            currentRoleCall.currentNodeID = Integer.MIN_VALUE;
+
+            discoverNewNodeDetails(getCirclePlusMAC());
+
+            logger.debug("Discovering nodes with role call on Circle+ ({})", getCirclePlusMAC());
+            roleCall(MIN_NODE_ID);
+            startDiscoveryWatchJob();
+        }
+    }
+
+    private MACAddress getCirclePlusMAC() {
+        return stickHandler != null ? stickHandler.getCirclePlusMAC() : null;
+    }
+
+    private ThingStatus getStickStatus() {
+        return stickHandler != null ? stickHandler.getThing().getStatus() : ThingStatus.UNKNOWN;
+    }
+
+    private void handleAnnounceAwakeRequest(AnnounceAwakeRequestMessage message) {
+        discoverNewNodeDetails(message.getMACAddress());
+    }
+
+    private void handleInformationResponse(InformationResponseMessage message) {
+        MACAddress mac = message.getMACAddress();
+        DiscoveredNode node = discoveredNodes.get(mac);
+        if (node != null) {
+            node.deviceType = message.getDeviceType();
+            PlugwiseUtils.updateProperties(node.properties, message);
+            if (node.isDataComplete()) {
+                thingDiscovered(createDiscoveryResult(node));
+                discoveredNodes.remove(mac);
+                logger.debug("Finished discovery of {} ({})", node.deviceType, mac);
+            }
+        } else {
+            logger.debug("Received information response for already discovered node ({})", mac);
+        }
+    }
+
+    @Override
+    public void handleReponseMessage(Message message) {
+        switch (message.getType()) {
+            case ANNOUNCE_AWAKE_REQUEST:
+                handleAnnounceAwakeRequest((AnnounceAwakeRequestMessage) message);
+                break;
+            case DEVICE_INFORMATION_RESPONSE:
+                handleInformationResponse((InformationResponseMessage) message);
+                break;
+            case DEVICE_ROLE_CALL_RESPONSE:
+                handleRoleCallResponse((RoleCallResponseMessage) message);
+                break;
+            default:
```

Comment:
Eclipse shows a warning when I remove the default case of a switch. I've added trace logging that is useful for debugging and also makes the default case more useful.